opentelemetry-sdk/src/opentelemetry/sdk/metrics/export/aggregate.py

```diff
@@ -0,0 +1,62 @@
+# Copyright 2019, OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+import threading
+
+
+class Aggregator(abc.ABC):
+    """Base class for aggregators.
+
+    Aggregators are responsible for holding aggregated values and taking a
+    snapshot of these values upon export (check_point).
+    """
+
+    def __init__(self):
+        self.current = None
+        self.check_point = None
+        self._lock = threading.Lock()
+
+    @abc.abstractmethod
+    def update(self, value):
+        """Updates the current with the new value."""
+
+    @abc.abstractmethod
+    def checkpoint(self):
+        """Stores a snapshot of the current value."""
+
+    @abc.abstractmethod
+    def merge(self, other):
+        """Combines two aggregator values."""
+
+
+class CounterAggregator(Aggregator):
+    """Aggregator for Counter metrics."""
+
+    def __init__(self):
+        super().__init__()
+        self.current = 0
+        self.check_point = 0
+
+    def update(self, value):
+        self.current += value
+
+    def checkpoint(self):
+        # TODO: Implement lock-free algorithm for concurrency
+        with self._lock:
+            self.check_point = self.current
+            self.current = 0
+
+    def merge(self, other):
+        self.check_point += other.check_point
```

Comment:
should current be updated as well? seems like this addition would be cleared immediately by the next call to checkpoint, which will match what exists in current.

Reply:
`merge` is usually called right after `checkpoint()` in the `process` step in a `Batcher`. This handles the race condition where an update to the metric happens at the same time as `process`. `merge` updates the `checkpoint` value to prepare for exporting, while the current is left untouched (but changed by the update seperately to be exported later on).